saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/OpenSamlAuthenticationProvider.java

```diff
@@ -211,6 +211,32 @@
 
 	private Converter<Saml2AuthenticationToken, Decrypter> decrypterConverter = new DecrypterConverter();
 
+	private Decrypter decrypter;
```

Comment:
This isn't thread-safe. Because the `decrypter` instance is changed with each authentication, two concurrent authentication attempts will clobber each other's value here.

Instead, the default values below will need to construct a `Decrypter` using the details from `Saml2AuthenticationToken`. For `assertionDecrypter`, I think this will look something like:

```java
Decrypter decrypter = this.decrypterConverter.convert(responseToken.getToken());
for (EncryptedAssertion encryptedAssertion : responseToken.getResponse().getEncryptedAssertions()) {
// ...
```

with something similar being done in `principalDecrypter`.